### DISCUSSION TITLE
Modular Forge Frontend

### DISCUSSION THREAD

**Discussion by armindarvish** [magit/forge#584] [Ideas]
Hi,

I am working on a package for browsing GitHub inside Emacs instead of using the browser, [consult-gh](https://github.com/armindarvish/consult-gh). I recently added support for forge (see [here](https://github.com/armindarvish/consult-gh#magitforge-integration)), using *forge* as the frontend to view issues. However, I had to hack everything to decouple forge from magit or create temp folders to trick magit to believe it is inside a git repository. I was wondering if there is any functions in forge (or if not is it possible to add some) that are more modular and do not heavily depend on Magit, Ghub, etc. Simply, I need a function that takes the name (or url) of a repo and an issue/pr number and I want to load it in a forge-topic buffer and then have the ability to add/edit posts, ... My current solution works, but it is hacky and what I am doing there really belongs in *forge* and not *consult-gh*! 

In general, I would like to have some modular capability to integrate *consult-gh* with *forge*. For example, I am working on the ability to switch between multiple accounts, but don't know how to support that with forge and ghub integration either. I'll file another issue in ghub for this specific one, though.

**Comment by tarsius** [comment]
Answering at https://github.com/magit/ghub/discussions/163 for now.

**Comment by tarsius** [comment]
> However, I had to hack everything to decouple forge from magit or create temp folders to trick magit to believe it is inside a git repository.

With Forge v0.4.0 such hackery should not be necessary anymore. Please remove that before `consult-gh` is added to Melpa (https://github.com/melpa/melpa/pull/9118).

Instead use `forge-topic-setup-buffer` and `forge-topics-setup-buffer`. Note how they don't need a Git repository and fall back to using `/` as `default-directory`.

**Reply by tarsius** [reply]
I.e., it boils down too

```elisp
(forge-topic-setup-buffer
 (forge-get-topic
  (forge-get-repository "https://github.com/magit/forge" nil :tracked)
  584))
```

Until a few minutes ago you would have had to use one of `forge-get{discussion,issue,pullreq}` instead, but that's now fixed with 243aa87d700a815434bf31ab66a9e4cf1159ac50.

**Reply by tarsius** [reply]
I've just added `forge-visit-topic-from-url`, which you may also find useful.
